Login and logout authentication with json-web-token

yujuncho opened the pull request:
This pull request adds api routes to login and create a user. Also included is the login and logout flow for front-end. The front-end uses localstorage to save the token and to determine if a user is logged in or not. Beyond these core features, the following changes were also made.

- Fixed a bug where npm i lead to an infinite loop
- Field components on the front-end are now configurable to show as a form-group or form-row
- The front-end was refactored to use one BrowserRouter, so the front-end no longer reloads when navigating between the landing page and /main
- Removed unneeded service-worker.ts in the public folder, and added functionality to cache font-awesome

yujuncho committed:
Created userRoutes and userController

yujuncho committed:
Implemented the api routes for /user/login and /user/signup

yujuncho committed:
Added an auth context, restructred routes such that there's no page reload, and created a skeleton login/logout flow

yujuncho committed:
Created login and signup pages and edited field-renderer and field such that fields can be configured as form-group or form-rows

yujuncho committed:
Fixed infinite loop with package.json npm i and updated the 'main'

yujuncho committed:
- Added navigation to the auth pages
- Updated the landing page's navigation to use HashLink
- Adjusted the margins for the auth page and the /main pages

yujuncho committed:
Implemented usage of tokens and localstorage to manage the logged in
status

- Added error handling and loading state for the Auth page
- Refactored Auth.js such that the form fields is maintained in a
separate file, auth-fields.ts
- Creaetd a useAuth hook to reduce the logic in App.tsx
- Updated the routes in App.tsx such that the landing page is accessible
for logged in users

yujuncho committed:
Added a reset logic for the Auth page when navigating between the signup
and login page. Also removed unnecessary escapes for the emailRegex for
the Auth Page's email input.

yujuncho committed:
Fixed a bug in the Auth page's error handling where the page was trying to render an error object rather than the error message

yujuncho committed:
Added loader animation for when the auth hook is checking local storage, and updated the service worker to cache fonts

yujuncho committed:
Removed unneeded service-worker.ts in the public folder

adbahrani commented on client/src/App.tsx:
```diff
@@ -1,18 +1,56 @@
+  const { userData, login, logout, checkedStorage } = useAuth();
+
+  if (!checkedStorage) {
+    return <Loader />;
```
with react we usually falsy and truthy values javascript. While thicicall this is correct and will works. More tipical way to right this is `!checkedStorage && <Loader />`

adbahrani commented on client/src/App.tsx:
```diff
@@ -1,18 +1,56 @@
+    return <Loader />;
+  }
+
+  let routes;
```
Typically we use <Router> tag and checkedStorage combine userData.token and && to enable routes, we can discuss that more

adbahrani commented on client/src/ehr_components/Main.tsx:
```diff
@@ -1,38 +1,29 @@
   const store = createStore(reducers);
-  let { path } = useRouteMatch();
+
+  const { path } = useRouteMatch();
```
Curious to see why you've used this

adbahrani commented on client/src/App.tsx:
```diff
@@ -1,18 +1,56 @@
+  }
+
+  let routes;
+  if (userData.token) {
```
Another way to do this is wrap all routes with AuthContextProvider
and then have  <PrivateRoute /> and <Route />

` export default function PrivateRoute({ component: Component, ...rest }) {
  const { isLoggedIn} = useAuth();

  return (
    <Route
      {...rest}
      render={props => {
        return isLoggedIn ? (
          <Component {...props} />
        ) : (
          <Redirect to="/login" />
        );
      }}
    ></Route>
  );
}`

adbahrani commented on client/src/components/header.tsx:
```diff
@@ -22,9 +24,9 @@ export const Header: React.FC<Props> = ({ data }) => {
                 </h1>
                 <p>{data.paragraph ? data.paragraph : "Loading.."}</p>
-                <a href="/main" className=" bttn-custom  ">
+                <Link to="/login" className=" bttn-custom  ">
```
👍👍

adbahrani commented on client/src/components/navigation.jsx:
```diff
@@ -1,69 +1,71 @@
+        isFixed ? "fixed-top" : ""
+      }`}
+    >
+      <Link className="navbar-brand page-scroll" to="/#page-top">
```
Curios to see why we change this to HashLink rather than just <a> , I believe you can ues NavLink as well

adbahrani commented on client/src/store/auth-context.tsx:
```diff
@@ -0,0 +1,11 @@
+import React from "react";
+
+export const AuthContext = React.createContext({
+  isLoggedIn: false,
```
having username would be also useful to have if we needed to display that, could be email as well.

adbahrani commented on models/UserModel.ts:
```diff
@@ -1,17 +1,15 @@
 import { Schema, model } from "mongoose";
 
-export interface User {
+export interface IUser {
```
👍

adbahrani commented on server.ts:
```diff
@@ -16,12 +21,13 @@ mongoose
     useFindAndModify: false,
-    useUnifiedTopology: true
+    useUnifiedTopology: true,
+    useCreateIndex: true
```
👍

yujuncho committed:
Fixed memory leak bug where Auth tried to update state after being
unmounted and search.tsx was rendered twice unnecessarily. The fix was
to stop updating state after calling authContext.login and to let
authContext.login handle the redirect rather than calling history.push.
Also made minor changes to...

- Add email to the stored userData
- Removed unnecessary comments and unused lines of code
